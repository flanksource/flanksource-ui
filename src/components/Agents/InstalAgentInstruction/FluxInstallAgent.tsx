import { JSONViewer } from "@flanksource-ui/ui/Code/JSONViewer";
import Handlebars from "handlebars";
import { useMemo } from "react";
import { TemplateContextData } from "./InstallAgentModal";

// This a  Handlebars template for the HelmRelease to install the agent and the
// kubernetes agent if the user has enabled it.
const fluxTemplate = `apiVersion: v1
kind: Namespace
metadata:
  name:  {{ namespace }}
---
{{#if createRepository}}
apiVersion: source.toolkit.fluxcd.io/v1beta1
kind: HelmRepository
metadata:
  name: {{ repoName }}
  namespace: {{ namespace }}
spec:
  interval: 5m0s
  url: https://flanksource.github.io/charts
---
{{/if}}
apiVersion: helm.toolkit.fluxcd.io/v2beta1
kind: HelmRelease
metadata:
  name: {{{ chart }}}
  namespace: {{ namespace }}
spec:
  chart:
    spec:
      chart: {{{ chart }}}
      sourceRef:
        kind: HelmRepository
        name: {{{ repoName }}}
        namespace: {{{ namespace }}}
      interval: 5m0s
  values:
{{#each values}}
    {{{ this }}}
{{/each}}

{{#if kubeValues}}
---
apiVersion: helm.toolkit.fluxcd.io/v2beta1
kind: HelmRelease
metadata:
  name: mission-control-kubernetes
  namespace: mission-control-agent
spec:
  chart:
    spec:
      chart: mission-control-kubernetes
      sourceRef:
        kind: HelmRepository
        name: flanksource
        namespace: mission-control-agent
  values:
  {{#each  kubeValues}} 
    {{{ this }}}
  {{/each}}
{{/if}}
  `;

const template = Handlebars.compile(fluxTemplate);

type Props = {
  data: TemplateContextData;
};

export default function FluxInstallAgent({ data }: Props) {
  const yaml = useMemo(() => {
    return template(data, {});
  }, [data]);

  return (
    <div className="flex max-h-[30rem] flex-1 flex-col gap-4 overflow-y-auto rounded-md border border-gray-200 p-2 px-4">
      <JSONViewer code={yaml} format="yaml" showLineNo />
    </div>
  );
}
