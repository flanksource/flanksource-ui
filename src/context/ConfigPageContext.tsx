import React, { createContext, useContext, useState } from "react";
import { ConfigItem } from "../api/types/configs";

export type ConfigState = {
  data?: ConfigItem[];
  filteredData?: ConfigItem[];
};

export type ConfigPageState = {
  configState: ConfigState;
  setConfigState: React.Dispatch<React.SetStateAction<ConfigState>>;
};

const initialState: ConfigPageState = {
  configState: {
    data: [],
    filteredData: []
  },
  setConfigState: ({ ...props }) => {}
};

const ConfigPageContext = createContext(initialState);

export const ConfigPageContextProvider = ({
  children
}: {
  children: React.ReactNode;
}) => {
  const [configState, setConfigState] = useState({
    ...initialState.configState
  });
  return (
    <ConfigPageContext.Provider value={{ configState, setConfigState }}>
      {children}
    </ConfigPageContext.Provider>
  );
};

export const useConfigPageContext = () => {
  const context = useContext(ConfigPageContext);

  if (context === undefined) {
    throw new Error(
      "useConfigPageContext must be used within a ConfigPageContextProvider"
    );
  }
  return context;
};
