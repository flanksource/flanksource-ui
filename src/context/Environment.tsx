export function IsAuthEnabled() {
  return process.env.NEXT_PUBLIC_WITHOUT_SESSION !== "true";
}

export function BaseUrl() {
  return process.env.BASE_URL;
}
