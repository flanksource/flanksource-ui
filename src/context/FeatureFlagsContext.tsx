import React, { createContext, useContext, useEffect, useState } from "react";
import { features } from "../services/permissions/features";
import {
  Property,
  permissionService
} from "../services/permissions/permissionsService";

export type FeatureFlagsState = {
  featureFlags: Property[];
  featureFlagsLoaded: boolean;
  refreshFeatureFlags: () => void;
  isFeatureDisabled: (_: keyof typeof features) => boolean;
};

const initialState: FeatureFlagsState = {
  featureFlags: [],
  featureFlagsLoaded: false,
  refreshFeatureFlags: () => {},
  isFeatureDisabled: (_) => false
};

const FeatureFlagsContext = createContext(initialState);

export const FeatureFlagsContextProvider = ({
  children
}: {
  children: React.ReactNode;
}) => {
  const [featureFlags, setFeatureFlags] = useState(initialState.featureFlags);
  const [featureFlagsLoaded, setFeatureFlagsLoaded] = useState(
    initialState.featureFlagsLoaded
  );

  const refreshFeatureFlags = async () => {
    const { data = [] } = await permissionService.loadProperties();
    setFeatureFlags(data!);
    setFeatureFlagsLoaded(true);
  };

  const isFeatureDisabled = (featureName: string) => {
    return permissionService.isFeatureDisabled(featureName, featureFlags);
  };

  useEffect(() => {
    refreshFeatureFlags();
  }, []);

  return (
    <FeatureFlagsContext.Provider
      value={{
        featureFlags,
        featureFlagsLoaded,
        refreshFeatureFlags,
        isFeatureDisabled
      }}
    >
      {children}
    </FeatureFlagsContext.Provider>
  );
};

export const useFeatureFlagsContext = () => {
  const context = useContext(FeatureFlagsContext);

  if (context === undefined) {
    throw new Error(
      "useFeatureFlagsContext must be used within a FeatureFlagsContextProvider"
    );
  }
  return context;
};
