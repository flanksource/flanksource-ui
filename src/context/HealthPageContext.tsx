import React, {
  Dispatch,
  SetStateAction,
  createContext,
  useContext,
  useState
} from "react";
import { URLSearchParamsInit } from "react-router-dom";
import { HealthCheck } from "../api/types/health";

export type HealthState = {
  checks: HealthCheck[] | null;
  filteredChecks: any[];
  filteredLabels: any;
  passing: {
    checks: number;
    filtered: number;
  };
  searchParams: URLSearchParamsInit;
};

export type HealthPageState = {
  healthState: HealthState;
  setHealthState: Dispatch<SetStateAction<HealthState>>;
};

const initialState: HealthPageState = {
  healthState: {
    checks: [],
    filteredChecks: [],
    filteredLabels: null,
    passing: {
      checks: 0,
      filtered: 0
    },
    searchParams: {}
  },
  setHealthState: ({ ...props }) => {}
};

const HealthPageContext = createContext(initialState);

export const HealthPageContextProvider = ({
  children
}: {
  children: React.ReactNode;
}) => {
  const [healthState, setHealthState] = useState({
    ...initialState.healthState
  });
  return (
    <HealthPageContext.Provider value={{ healthState, setHealthState }}>
      {children}
    </HealthPageContext.Provider>
  );
};

export const useHealthPageContext = () => {
  const context = useContext(HealthPageContext);

  if (context === undefined) {
    throw new Error(
      "useHealthPageContext must be used within a HealthPageContextProvider"
    );
  }
  return context;
};
