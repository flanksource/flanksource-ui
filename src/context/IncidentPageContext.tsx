import React, { createContext, useContext, useState } from "react";

export type IncidentState = {
  incidents: any[] | null;
  ownerSelections: any[] | null;
};

export type IncidentPageState = {
  incidentState: IncidentState;
  setIncidentState: ({ ...props }: IncidentState) => any;
};

const initialState: IncidentPageState = {
  incidentState: {
    incidents: [],
    ownerSelections: []
  },
  setIncidentState: ({ ...props }) => {}
};

const IncidentPageContext = createContext(initialState);

export const IncidentPageContextProvider = ({
  children
}: {
  children: React.ReactNode;
}) => {
  const [incidentState, setIncidentState] = useState({
    ...initialState.incidentState
  });
  return (
    <IncidentPageContext.Provider value={{ incidentState, setIncidentState }}>
      {children}
    </IncidentPageContext.Provider>
  );
};

export const useIncidentPageContext = () => {
  const context = useContext(IncidentPageContext);

  if (context === undefined) {
    throw new Error(
      "useIncidentPageContext must be used within a IncidentPageContextProvider"
    );
  }
  return context;
};
