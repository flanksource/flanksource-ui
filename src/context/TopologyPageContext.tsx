import React, { useState, createContext, useContext } from "react";
import { URLSearchParamsInit } from "react-router-dom";

export type ValueType = number | string | Date;

export type TopologyProperty = {
  name: string;
  icon?: string;
  type?: string;
  text?: string;
  max?: number;
  min?: number;
  headline?: boolean;
  value?: ValueType;
  unit?: string;
};

export type Topology = {
  id: string;
  name: string;
  type?: string;
  updated_at?: string;
  properties?: TopologyProperty[];
};

export type TopologyState = {
  topology: Topology[] | null;
  searchParams: URLSearchParamsInit;
};

export type TopologyPageState = {
  topologyState: TopologyState;
  setTopologyState: ({ ...props }: TopologyState) => any;
};

const initialState: TopologyPageState = {
  topologyState: {
    topology: null,
    searchParams: {}
  },
  setTopologyState: ({ ...props }) => {}
};

const TopologyPageContext = createContext(initialState);

export const TopologyPageContextProvider = ({
  children
}: {
  children: React.ReactElement | React.ReactElement[];
}) => {
  const [topologyState, setTopologyState] = useState({
    ...initialState.topologyState
  });
  return (
    <TopologyPageContext.Provider value={{ topologyState, setTopologyState }}>
      {children}
    </TopologyPageContext.Provider>
  );
};

export const useTopologyPageContext = () => {
  const context = useContext(TopologyPageContext);

  if (context === undefined) {
    throw new Error(
      "useTopologyPageContext must be used within a TopologyPageContextProvider"
    );
  }
  return context;
};
