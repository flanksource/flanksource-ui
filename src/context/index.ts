import { createContext, useContext } from "react";
import { User } from "../api/types/users";
import { Authorizer as casbinAuthorizer } from "casbin.js";
import { permDefs } from "./UserAccessContext/permissions";
import { WhoamiResponse } from "@flanksource-ui/api/services/users";
export type ActionType = "write" | "read";

export const Roles = {
  admin: "admin",
  editor: "editor",
  viewer: "viewer",
  commander: "commander",
  responder: "responder"
};

export function createAuthorizer(ctx: WhoamiResponse["payload"]): Authorizer {
  const authorizer = new casbinAuthorizer("manual");
  const builtPerms: Record<string, any> = {};

  ctx.roles.forEach((role) => {
    const permissions = permDefs[role as keyof typeof permDefs] ?? {};
    Object.entries(permissions).forEach(([key, value]) => {
      builtPerms[key] = [...(builtPerms[key] || []), ...value];
    });
  });
  authorizer.setPermission(builtPerms);
  return {
    hasResourceAccess: (resourceName, action) => {
      return authorizer.can(action, resourceName);
    },
    hasAnyResourceAccess: (resourceNames, action) => {
      return authorizer.canAny(action, resourceNames);
    }
  };
}

export type Authorizer = {
  hasResourceAccess: (
    resourceName: string,
    action: ActionType
  ) => Promise<boolean>;
  hasAnyResourceAccess: (
    resourceNames: string[],
    action: ActionType
  ) => Promise<boolean>;
};

export type Permission = {
  subject: string;
  object: string;
  deny?: boolean;
  action: string;
};

interface IAuthContext {
  user?: User;
  backendUrl?: string;
  orgSlug?: string;
  roles: string[];
  permissions: Permission[];
  isLoading?: boolean;
  authorizer: Authorizer;
}

export const AuthContext = createContext<IAuthContext>({
  roles: [],
  permissions: [],
  isLoading: false,
  authorizer: {
    hasResourceAccess: () => Promise.resolve(false),
    hasAnyResourceAccess: () => Promise.resolve(false)
  }
});

export const useUser = () => useContext(AuthContext);

export const FakeUser = (roles: string[]): IAuthContext => {
  let user = {
    id: "b149b5ee-db1c-4c0c-9711-98d06f1f1ce7",
    email: "admin@local",
    name: "John Doe"
  };
  return {
    roles: roles,
    user: user,
    permissions: [],
    isLoading: false,
    backendUrl: "https://testurl.com",
    authorizer: createAuthorizer({
      roles: roles,
      database: "",
      hostname: "",
      permissions: [],
      user: user
    })
  };
};
