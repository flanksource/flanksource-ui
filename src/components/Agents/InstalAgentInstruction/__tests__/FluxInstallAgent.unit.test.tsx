import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import { AuthContext, FakeUser, Roles } from "../../../../context";
import FluxInstallAgent from "../FluxInstallAgent";

global.ResizeObserver = jest.fn().mockImplementation(() => ({
  observe: jest.fn(),
  unobserve: jest.fn(),
  disconnect: jest.fn()
}));

const writeText = jest.fn();

const mockInput = {
  chart: "mission-control-agent",
  namespace: "mission-control-agent",
  repoName: "flanksource",
  values: [
    "upstream.createSecret=true",
    "upstream.host=http://localhost:3000",
    "upstream.username=token",
    "upstream.password=password",
    "upstream.agentName=test-new-agent-instructions",
    "pushTelemetry.enabled=true",
    "pushTelemetry.topologyName=https://incident-commander.demo.aws.flanksource.com-test-new-agent-instructions"
  ]
};

const mockInputWithKubOptions = {
  chart: "mission-control-agent",
  namespace: "mission-control-agent",
  repoName: "flanksource",
  values: [
    "upstream.createSecret=true",
    "upstream.host=http://localhost:3000",
    "upstream.username=token",
    "upstream.password=password",
    "upstream.agentName=test-new-agent-instructions",
    "pushTelemetry.enabled=true",
    "pushTelemetry.topologyName=https://incident-commander.demo.aws.flanksource.com-test-new-agent-instructions"
  ],
  kubeValues: [
    'clusterName: "test-new-agent-instructions"',
    'scraper.schedule: "30m"'
  ]
};

Object.assign(navigator, {
  clipboard: {
    writeText
  }
});

describe("InstallAgentModal", () => {
  it("renders the Helm repository installation command", async () => {
    render(
      <AuthContext.Provider value={FakeUser([Roles.admin])}>
        <FluxInstallAgent
          agentFormValues={{
            name: "testname",
            kubernetes: {},
            properties: {}
          }}
          generatedAgent={generatedAgent}
        />
      </AuthContext.Provider>
    );

    const btn = screen.getByTitle(/Copy to clipboard/i);

    fireEvent.click(btn);

    await waitFor(() => {
      expect(writeText).toHaveBeenCalled();
    });
    expect(writeText.mock["calls"][0][0]).toMatchSnapshot();
  });

  it("renders the Helm repository installation command with kube command", async () => {
    render(
      <AuthContext.Provider value={FakeUser([Roles.admin])}>
        <FluxInstallAgent
          generatedAgent={generatedAgent}
          agentFormValues={{
            name: "testname",
            properties: {},
            kubernetes: {
              interval: "1m",
              exclusions: ["testexclusion1", "testexclusion2"]
            }
          }}
        />
      </AuthContext.Provider>
    );
    const btn = screen.getByTitle(/Copy to clipboard/i);

    fireEvent.click(btn);

    await waitFor(() => {
      expect(writeText).toHaveBeenCalled();
    });
    expect(writeText.mock["calls"][0][0]).toMatchSnapshot();
  });
});
