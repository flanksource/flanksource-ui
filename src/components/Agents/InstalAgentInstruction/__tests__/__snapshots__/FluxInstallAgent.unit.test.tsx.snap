// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InstallAgentModal renders the Helm repository installation command 1`] = `
"apiVersion: v1
kind: Namespace
metadata:
  name:  mission-control-agent
---
apiVersion: helm.toolkit.fluxcd.io/v2beta1
kind: HelmRelease
metadata:
  name: mission-control-agent
  namespace: mission-control-agent
spec:
  chart:
    spec:
      chart: mission-control-agent
      sourceRef:
        kind: HelmRepository
        name: flanksource
        namespace: mission-control-agent
      interval: 5m0s
  values:
    upstream.createSecret=true
    upstream.host=http://localhost:3000
    upstream.username=token
    upstream.password=password
    upstream.agentName=test-new-agent-instructions
    pushTelemetry.enabled=true
    pushTelemetry.topologyName=https://incident-commander.demo.aws.flanksource.com-test-new-agent-instructions

  "
`;

exports[`InstallAgentModal renders the Helm repository installation command with kube command 1`] = `
"apiVersion: v1
kind: Namespace
metadata:
  name:  mission-control-agent
---
apiVersion: helm.toolkit.fluxcd.io/v2beta1
kind: HelmRelease
metadata:
  name: mission-control-agent
  namespace: mission-control-agent
spec:
  chart:
    spec:
      chart: mission-control-agent
      sourceRef:
        kind: HelmRepository
        name: flanksource
        namespace: mission-control-agent
      interval: 5m0s
  values:
    upstream.createSecret=true
    upstream.host=http://localhost:3000
    upstream.username=token
    upstream.password=password
    upstream.agentName=test-new-agent-instructions
    pushTelemetry.enabled=true
    pushTelemetry.topologyName=https://incident-commander.demo.aws.flanksource.com-test-new-agent-instructions

  "
`;
